tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_on_block.py

```diff
@@ -0,0 +1,691 @@
+    #         == fork_3_state.current_justified_checkpoint
+    #         != store.best_justified_checkpoint)
+    # assert (store.best_justified_checkpoint
+    #         == fork_2_state.current_justified_checkpoint)
+
+    yield 'steps', test_steps
+
+
+@with_all_phases
+@spec_state_test
+def test_new_finalized_slot_is_not_justified_checkpoint_ancestor(spec, state):
+    """
+    J: Justified
+    F: Finalized
+    state (forked from genesis):
+        epoch
+        [0] <- [1] <- [2] <- [3] <- [4] <- [5]
+         F                    J
+
+    another_state (forked from epoch 0):
+         └──── [1] <- [2] <- [3] <- [4] <- [5]
+                       F      J
+    """
+    test_steps = []
+    # Initialization
+    store, anchor_block = get_genesis_forkchoice_store_and_block(spec, state)
+    yield 'anchor_state', state
+    yield 'anchor_block', anchor_block
+    current_time = state.slot * spec.config.SECONDS_PER_SLOT + store.genesis_time
+    on_tick_and_append_step(spec, store, current_time, test_steps)
+    assert store.time == current_time
+
+    # ----- Process state
+    # Goal: make `store.finalized_checkpoint.epoch == 0` and `store.justified_checkpoint.epoch == 3`
+    # Skip epoch 0
+    next_epoch(spec, state)
+
+    # Forking another_state
+    another_state = state.copy()
+
+    # Fill epoch 1 with previous epoch attestations
+    state, store, _ = yield from apply_next_epoch_with_attestations(
+        spec, state, store, False, True, test_steps=test_steps)
+    # Skip epoch 2
+    next_epoch(spec, state)
+    # Fill epoch 3 & 4 with previous epoch attestations
+    for _ in range(2):
+        state, store, _ = yield from apply_next_epoch_with_attestations(
+            spec, state, store, False, True, test_steps=test_steps)
+
+    assert state.finalized_checkpoint.epoch == store.finalized_checkpoint.epoch == 0
+    assert state.current_justified_checkpoint.epoch == store.justified_checkpoint.epoch == 3
+    assert store.justified_checkpoint == state.current_justified_checkpoint
+
+    # Create another chain
+    # Goal: make `another_state.finalized_checkpoint.epoch == 2` and `another_state.justified_checkpoint.epoch == 3`
+    all_blocks = []
+    # Fill epoch 1 & 2 with previous + current epoch attestations
+    for _ in range(3):
+        _, signed_blocks, another_state = next_epoch_with_attestations(spec, another_state, True, True)
+        all_blocks += signed_blocks
+
+    assert another_state.finalized_checkpoint.epoch == 2
+    assert another_state.current_justified_checkpoint.epoch == 3
+    assert state.finalized_checkpoint.hash_tree_root() != another_state.finalized_checkpoint.hash_tree_root()
+    assert (
+        state.current_justified_checkpoint.hash_tree_root()
+        != another_state.current_justified_checkpoint.hash_tree_root()
+    )
+    # pre_store_justified_checkpoint_root = store.justified_checkpoint.root
+
+    # FIXME: pending on the `on_block`, `on_attestation` fix
```

Comment:
ditto, as `test_new_justified_is_later_than_store_justified`, it hits the `store.justified_checkpoint is not in store.checkpoint_states` issue.